modules/generator/registry/counter.go

```diff
@@ -121,12 +137,26 @@ func (c *counter) collectMetrics(appender storage.Appender, timeMs int64, extern
 	}
 
 	for _, s := range c.series {
+		t := time.UnixMilli(timeMs)
 		// set series-specific labels
 		for i, name := range c.labels {
 			lb.Set(name, s.labelValues[i])
 		}
 
-		_, err = appender.Append(0, lb.Labels(nil), timeMs, s.value.Load())
+		// If we are about to call Append for the first time on a series, we need
+		// to first insert a 0 value to allow Prometheus to start from a non-null
+		// value.
+		if s.isNew() {
+			_, err = appender.Append(0, lb.Labels(nil), timeMs, 0)
+			if err != nil {
+				return
+			}
+			// Increment timeMs to ensure that the next value is not at the same time.
+			t = t.Add(insertOffsetDuration)
+		}
+
+		s.registerSeenSeries()
```

Comment:
Nit: if we move this within the if case we only execute this when we actually have to change the value. I can't imagine the overhead of this will be big but still, no need to keep writing false.

Reply:
Good call, I'll move.